mlir/include/mlir/IR/MLIRContext.h

```diff
@@ -240,6 +241,9 @@ class MLIRContext {
   /// (attributes, operations, types, etc.).
   llvm::hash_code getRegistryHash();
 
+  /// Get context-specific system description
+  SystemDesc &getSystemDesc();
```

Comment:
This can't belong the MLIRContext in any way: the context can be used in parallel for compiling to different targets.

This likely should be an attribute in the IR instead.

Reply:
We discussed this internally and there isn't a clear answer. We do agree that we need an IR annotation, that's why the target descriptor has a `targetID`. In theory, the context is only being used as a singleton for the whole module, and the way you query the target is to get the operation's target attribute (or its parents) and query the context for a target of that ID. This is not the most robust design, but we're still grasping at straws on how this will all fit together, so it sounded like a reasonable starting point.

There is usually a single host there could be multiple targets (including the host), but the host can be a different target than the one you're compiling on. For example, compile in the head node (older machine) to run on a cluster node (newer machine with multiple devices). So we need to provide at least one host and one target. Of course, the default value for those is the same: the current machine. Passing "names" would trigger an LLVM target lookup, passing config files would trigger parsing of such files. We could treat host/target as just target(s), not sure we need a difference here.

In the IR, we may use different target descriptors for different functions, or perhaps even different ops (say, a single `linalg.generic`), but that granularity can be too costly. So we're discussing if we leave the target information as a function attribute (and outline target code) or if we create regions that define targets, like the ones proposed by another team we collaborate with (@fschlimb, @Hardcode84).

To avoid this discussion, we just have here the possibility of having multiple devices, but not yet making full use of it. We're trying to get the basic functionality first before we start discussing the merits of target annotations and where they live (and how they're recursively searched if not found, and the compile time cost of it).

Reply:
Just to expand on the target lookup strategy, this is one alternative we discussed internally. Ignore the `getContext()` for now, let's just assume this is whatever gets you the target descriptor handler.

The first (naive and wrong) behaviour could be:
* If the op has a target attribute:
  * `target = op.getContext().getTarget(op.getTargetID()); // This could be an ID or a string into a map`
* Else
  * `target = op.getContext().getDefaultTarget(); // Default target is target 0`

A more advanced lookup would need to recurse:
* Else
  * Iterate through parents until target ID is found (module should be target 0)
  * `target = op.getContext().getTarget(op.getTargetID());`

This would be costly, as no op today has a target ID and we assume less of them will have it than not have it.

Two alternatives are:
1. Mandate every op to have a target ID. This breaks compatibility, increases parsing and analysis costs, touches every op, including downstream ones. Not viable.
2. Make so that only a few ops will ever have such attribute, a bit like a symbol table. This could be modules, functions and certain grouping operations (like Frank's / Ivan's). This is backwards compatible and you only pay the cost if you want a target descriptor in the first place.

The latter above is more complex than just taking the local op target, but it's a manageable complexity, unlike the other options. For example, a pass that uses target information and cost models would look for such operations and dispatch a transform / rewrite onto that region and nowhere else, passing the target descriptor with it, guaranteeing no recursive search will need to happen.

But any of these better solutions are into the future. Right now, we're worried about the basic infrastructure, for example:
* Where does it live? If not context, where would be a more appropriate global context that only spans the module?
* How do we query it? If not via device ID, do we just attach references to the actual target/model directly to the ops (regions, functions, modules)?

Also remember, the compiler isn't meant to use the target descriptor directly. This is an interim stage while we get the cost model up, and then passes/transforms will use the cost model to ask questions, which in turn will enquire the target descriptor for information.

Reply:
Also, another point that we thought of was not to increase size of static data by keeping `SystemDesc` as a global object. But if that is not a concern, then conceptually, a global `SystemDesc` seems right design. With that, any target can query properties of any other device/target, without being restrictive of the execution environment.

Reply:
> This would be costly, as no op today has a target ID and we assume less of them will have it than not have it.

I don't follow, this is something you do once at the beginning of the pass, and then cache and pass around.
Have you looked at the GPU target attributes? Or the DLTI?

Reply:
I think we're talking about different things. But I have mixed them up a bit on my answer above, so I understand the confusion.

One thing is op attribute to indicate target features that are specific to the op, generally low level ops like in the GPU dialect. Here, there's a pass that lowers IR into those ops with a particular semantics, encoded in the attributes.

Another is how do we get to those decisions in the first place. We may very well end up with a lot of target attributes encoded in IR, at high and low level, but there still needs to be a framework that will help find _which of them_ to create, or even how to tile and fuse more than one op, etc. This cannot be an attribute to a particular op.

In my answer I should have kept to the latter: where do we take those properties from (TTI, Json, cmd-line args); But I have also pushed a little bit on how to represent them (function/op attributes, regions). This PR is about the TTI infrastructure, not the IR representation.

Why did I mix them? Because of the `MLIRContext` question. As @nhasabni said, regardless of where we keep the TTI/Cost model, we will need access to it from passes, transforms, lowering, perhaps even the front-end will need to use it, in order to generate the correct IR ops or perform the most accurate analysis. 

Moreover, device transformations may need to know about the host properties to make efficient decisions, so we should have a way to ask about _any_ target from _any_ place in the code that handles IR. The `MLIRContext` is one of those places, but not the only one. It could even be a global structure.

We have discussed two designs:
1. A single target descriptor framework for all targets, indexed by some ID (numeric, text, hash) which is accessible from any op (like `op.getContext().getTTI()`).
2. A initial target descriptor framework for all targets that creates IR passing individual TTIs (as a pointer to a singleton) to each sub-graph (module? function? op?).

The first one seems simpler to implement and more generic to use. Since it's not clear what the actual core usage will be, we picked this one. Also, if you squint hard enough, you can see that (2) is basically the same as (1) in complexity, just a different API and harder to access _other targets'_ info.

Reply:
Sorry I don't quite follow, so I think I need to ask more questions:

>  This cannot be an attribute to a particular op

Why? What about the top-level module for example? The module could give you the handle to the entire target description (since this is repeating my previous pointer to DLTI and the GPU target ops, it likely means we're talking past each others here).

> Moreover, device transformations may need to know about the host properties to make efficient decisions, so we should have a way to ask about any target from any place in the code that handles IR. The MLIRContext is one of those places, but not the only one. 

I mean sure, but I still don't quite get the limitation of "an attribute on the ModuleOp": an attribute is in the MLIRContext already and you get access to it from the IR.

Reply:
> Most of the above gets clumsy if I have to edit the IR, remove some attributes (and not others), add some attributes (and have merge semantics), etc. And in the end, how many attributes will we have for any small piece of IR for any particular target? If LLVM's TTI is any indication, probably hundreds, per target.

I wouldn't suggest serializing all of TTI as attribute, the attribute is just a "target id" that implement the right interface. It's orthogonal to how the interface is implemented.

> How to use them as cost modelling, how to interpret different target attributes, and what do these things mean. There has to be a piece of infrastructure that handles all of this in MLIR regardless of where it lives or where it comes from. That infrastructure is what this PR is trying to build.

We probably need a better design doc, this PR as it is does not tell enough of the story for me to see where this is going.

Reply:
> where does that target id live - op/func/region/module - and how does that affect passes?

Isn't this something you have to solve anyway? We've been through this with DLTI already, I'm not sure about a better alternative, but it has to be in the IR somehow IMO (when you have a gpu module nested in a host module, they have two different targets, and two gpu modules with two different targets can be present as well).

> who populates the target id as attributes? How long do they live? where does the actual map lives?

I don't quite follow the question, mostly because DLTI or GPU targets interfaces are design point that already have answers for this. Meaning if you have an interface, anyone can provide it, the actual instance lives in the context (as every attribute do) in some fashion (not as a new first class concepts, but using all the existing storage we already have).

> do we want the maps to be a global? context? passed through as args?

I don't see how a globals make sense, we preserved MLIR from global state as much as possible and I'm skeptical we need to introduce some. The Context is a possibility for the storage, but as mentioned above it is already extensible: I don't expect any modification to the MLIRContext at the moment.
"passed as args" isn't an infra question: it become a tooling question. That is "at the edge" you can have something that translate some args into the initialized state (this is how we handle a lot of the options), but the infra never actually **need** any cl::opt parsing for setting up the infra.

> This problem is so large that it's virtually impossible to have a PR with a running example that covers all cases. 

This is not what I'm asking, I'm asking for a doc (not code) and not something that covers all cases, but just one single case that we can play with "on paper". Like starts simple: let's say we have a simple TOSA model you want to codegen for a particular modern CPU: we can make up some conceptual stages of the lowering, and look at some of the interactions between these lowering and a hypothetical target interface.

Reply:
Ok, fair enough. I think there's a lot to unpack and I feel we're mostly agreeing with each other but seeing different facets of the same (very large) picture.

A document is definitely a more reasonable approach than comments on a PR. This thread has definitely run its course. :)

We'll start some public document and update here. We'll continue from there. Thanks for the patience so far!! 😅